syntax = "proto3";

package redtangle;

// Remote API
service RedTangle {
    rpc Connect(Username) returns (ConnectionResponse) {}
    rpc GetStatus(Username) returns (Status) {}
    rpc Select(SelectRequest) returns (Status) {}
    rpc Rotate(RotateRequest) returns (Status) {}
    rpc EndTurn(Username) returns (Status) {}
    rpc Disconnect(Username) returns (StatusCode) {}
}

message Username {
    string username = 1;
}

message GameStatus {
    string opponent = 1;
    string turn = 2;
    string winner = 3;
    string board = 4;
}

message StatusCode {
    StatusCodes code = 1;
}

enum StatusCodes {
    NO_ERROR = 0;
    USER_EXISTS = 1;
    GAME_FULL = 2;
    SERVER_ERROR = 3;
}

message Status {
    GameStatus game_status = 1;
    StatusCodes server_status = 2;
}

message ConnectionResponse {
    string team_color = 1;
    Status status = 2;
}

message Position {
    int32 x = 1;
    int32 y = 2;
}

message SelectRequest {
    Username username = 1;
    Position position = 2;
}

message RotateRequest {
    Username username = 1;
    bool clockwise_rotation = 2;
}

message Empty {

}
